xds/internal/balancer/cdsbalancer/cluster_watcher.go

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2023 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cdsbalancer
+
+import (
+	"context"
+
+	"google.golang.org/grpc/xds/internal/xdsclient/xdsresource"
+)
+
+// clusterWatcher implements the xdsresource.ClusterWatcher interface, and is
+// passed to the xDS client as part of the WatchResource() API.
+//
+// It watches a single cluster and handles callbacks from the xDS client by
+// scheduling them on the parent LB policy's serializer.
+type clusterWatcher struct {
+	name   string
+	parent *cdsBalancer
+}
+
+func (cw *clusterWatcher) OnUpdate(u *xdsresource.ClusterResourceData) {
+	cw.parent.serializer.Schedule(func(context.Context) {
+		cw.parent.onClusterUpdate(cw.name, u.Resource)
+	})
+}
+
+func (cw *clusterWatcher) OnError(err error) {
+	cw.parent.serializer.Schedule(func(context.Context) {
+		cw.parent.onClusterError(cw.name, err)
+	})
+}
+
+func (cw *clusterWatcher) OnResourceDoesNotExist() {
+	cw.parent.serializer.Schedule(func(context.Context) {
+		cw.parent.onClusterResourceNotFound(cw.name)
+	})
+}
+
+// watcherState groups the state associated with a clusterWatcher.
+type watcherState struct {
+	watcher     *clusterWatcher            // The underlying watcher.
+	cancelWatch func()                     // Cancel func to cancel the watch.
+	lastUpdate  *xdsresource.ClusterUpdate // Most recent update received for this cluster.
```

Comment:
Optional: mention access to this is protected by callback serializer? This is from a higher level though so up to you.

Reply:
Yeah, this struct is simply a container to group a bunch of state. It is up to the user of this struct to arrange for appropriate synchronization.

Reply:
This type, specifically, doesn't provide any functionality other than being a container to hold a bunch of state. But any other type which holds some state and provides some functionality (that possibly uses that state) should document whether it guarantees synchronization and in what way, e.g. it is safe to invoke any of the methods concurrently, or it is safe to invoke methods a,b, and c concurrently but they cannot be invoked concurrently with d, e, f. See our `ClientStream`, `ServerStream` types for examples. Or, a type could simply say that it is not safe for concurrent access, and it is up to the callers to ensure appropriate synchronization.

I would have added something like that if this type was anything more than just a container of some state used by a single entity.

> In this specific case, I found it interesting that only lastUpdate had synchronization requirements of the callback serializer (other two pieces of state are only written at build time), which is the reason I wrote this comment.

That is because the other two are initialized when the `watcherState` is created, and are read-only after that point. The `lastUpdate` field is the only one which can change later on.